object-store/src/main/java/org/hypertrace/config/objectstore/IdentifiedObjectStore.java

```diff
@@ -153,24 +154,39 @@ public Optional<T> getData(RequestContext context, String id) {
     return getObject(context, id).map(ConfigObject::getData);
   }
 
-  public ContextualConfigObject<T> upsertObject(RequestContext context, T data) {
+  private ContextualConfigObject<T> upsertObject(
+      RequestContext context, UpsertConfigRequest request) {
     UpsertConfigResponse response =
         context.call(
-            () ->
-                this.configServiceBlockingStub
-                    .withDeadline(getDeadline())
-                    .upsertConfig(
-                        UpsertConfigRequest.newBuilder()
-                            .setResourceName(this.resourceName)
-                            .setResourceNamespace(this.resourceNamespace)
-                            .setContext(this.getContextFromData(data))
-                            .setConfig(this.buildValueFromData(data))
-                            .build()));
+            () -> this.configServiceBlockingStub.withDeadline(getDeadline()).upsertConfig(request));
 
     return this.processUpsertResult(context, response)
         .orElseThrow(Status.INTERNAL::asRuntimeException);
   }
 
+  public ContextualConfigObject<T> upsertObject(RequestContext context, T data) {
+    return this.upsertObject(
+        context,
+        UpsertConfigRequest.newBuilder()
+            .setResourceName(this.resourceName)
+            .setResourceNamespace(this.resourceNamespace)
+            .setContext(this.getContextFromData(data))
+            .setConfig(this.buildValueFromData(data))
+            .build());
+  }
+
+  public ContextualConfigObject<T> upsertObject(RequestContext context, T data, Filter filter) {
```

Comment:
I thought on this again.

We don't need new abstraction we can just overload the method in base store reasons:
1. The Filter is a generic class can be used directly without a generic type.
2. It was difficult to convert different type of request (like update, or get ) under a single abstract method which builds filter
3. we can build the filter in Impl class and pass the filter.
4. We can have overload on other get methods and this seems simpler for impl class also

Reply:
update dashboard will look like
```
@Override
  public void updateDashboard(
      UpdateDashboardRequest request, StreamObserver<UpdateDashboardResponse> responseObserver) {
    RequestContext requestContext = RequestContext.CURRENT.get();
    try {
      this.validator.validate(requestContext, request);
      Dashboard existingDashboard =
          this.dashboardStore
              .getData(requestContext, request.getId())
              .orElseThrow(Status.NOT_FOUND::asRuntimeException);

      Dashboard.Builder dashboardBuilder =
          existingDashboard.toBuilder()
              .setName(request.getName())
              .setJson(request.getJson())
              .setLastUpdatedTimestamp(timestampConverter.convert(clock.instant()));

      if (request.hasDescription()) {
        dashboardBuilder.setDescription(request.getDescription());
      }

      Filter upsertCondition =
          Filter.newBuilder()
              .setRelationalFilter(
                  RelationalFilter.newBuilder()
                      .setConfigJsonPath(
                          Dashboard.getDescriptor()
                              .findFieldByNumber(Dashboard.LAST_UPDATED_TIMESTAMP_FIELD_NUMBER)
                              .getJsonName())
                      .setOperator(RelationalOperator.RELATIONAL_OPERATOR_EQ)
                      .setValue(Values.of(Timestamps.toString(request.getPreviousUpdateTimestamp()))))
              .build();

      responseObserver.onNext(
          UpdateDashboardResponse.newBuilder()
              .setDashboard(
                  this.dashboardStore
                      .upsertObject(requestContext, dashboardBuilder.build(), upsertCondition)
                      .getData())
              .build());
      responseObserver.onCompleted();
    } catch (Exception exception) {
      Exception decoratedException = decorateException(requestContext, exception);
      log.warn(
          "Failed to update dashboard for request: {} with context: {}",
          request,
          requestContext,
          decoratedException);
      responseObserver.onError(exception);
    }
  }
```

Reply:
i get an idea but not exactly.
Let me try and push few commits for reference

Reply:
> with a few caveats
what are the limitation you see? i will try to see if i can handle them

Reply:
> what are the limitations you see? i will try to see if i can handle them

Just the points from the other comments. Mainly that 
- we should avoid writing custom conversion code - we can definitely figure something out between the existing converter and/or object mapper.
- the point about making the update w/ filter protected. We can let the impl class deal with building it and expose a typed api. We don't need the impl class exposing the generic filter method without the conversion (i.e. callers to the specific store shouldn't be building a filter)

Reply:
Made the above changes and corresponding config service code: https://github.com/Traceableai/config-service/pull/2055.

> I also think it could be worthwhile to provide some filter building utilities for easily going from a first class proto filter object to a generic one, but we can park that and implement it a few times before pulling out commonalities

on this what type of utilities i can add?

Reply:
Conversion looks good. Can you fix the second point - make your new method protected so it doesn't expose it to all callers.

> on this what type of utilities i can add?

Let's come back to this once we build a few implementations. But basically what I was thinking is most typed filters are a series of optional (field, primitive) pairs that are used with equality, and then we AND together any provided pairs. I suspect we could automatically build the generic filter with minimal config (e.g. providing a mapping of filter field number to the json path). But anyway as I said, we can revisit in the future only if it seems worthwhile.

Reply:
right, i know what can we add but at this stage not have enough sets to build one.